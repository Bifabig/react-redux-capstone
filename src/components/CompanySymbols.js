import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { NavLink, useNavigate } from 'react-router-dom';
import { getSlicedData, getSymbols } from '../redux/symbols/symbolsSlice';
import styles from '../styles/CompanySymbols.module.css';

const CompanySymbols = () => {
  const navigate = useNavigate();
  const [searchVal, setSearchVal] = useState('');
  const [fetchMax, setFetchMax] = useState(10);
  const [isSubmitted, setIsSubmitted] = useState(false);
  const dispatch = useDispatch();
  const {
    slicedSymbols, symbols, isLoading, error,
  } = useSelector(
    (state) => state.symbols,
  );
  useEffect(() => {
    if (symbols.length === 0) {
      dispatch(getSymbols());
    }
  }, [dispatch, symbols]);

  const handleSubmit = (e) => {
    e.preventDefault();
    setIsSubmitted(true);
    return navigate('/search', { state: { searchVal, isSubmitted } });
  };

  const fetchData = () => {
    if (symbols.length !== slicedSymbols.length) {
      dispatch(getSlicedData(fetchMax));
      setFetchMax(fetchMax + 30);
    } else {
      setFetchMax(-1);
    }
  };

  if (error) return <h2>Something went wrong</h2>;

  return isLoading ? <div className={styles.container}><h2>Loading...</h2></div> : (
    <div className={styles.container}>
      <div className={styles.input}>

        <form onSubmit={handleSubmit}>
          <input
            type="text"
            placeholder="Search..."
            value={searchVal}
            onChange={(e) => setSearchVal(e.target.value)}
          />

          <input
            type="submit"
            value="search"
          />

        </form>
      </div>
      <div className={styles.stocks}>

        <h2>
          <i className="bi bi-bar-chart-steps" />
          Stocks
        </h2>

        <div
          className={styles.symbols}
        >

          {slicedSymbols.map((symbol) => (
            <NavLink to="/balanceSheets" key={symbol} state={{ symbol: { symbol } }}>
              <li>

                {symbol}
                {' '}
                <i className="bi bi-arrow-right-circle" />
              </li>
            </NavLink>
          ))}
          <div className={styles.btnContainer}>
            {fetchMax !== -1 ? <button type="button" onClick={fetchData} className={styles.load}>Load More</button> : ''}

          </div>
        </div>

      </div>
    </div>
  );
};

export default CompanySymbols;
