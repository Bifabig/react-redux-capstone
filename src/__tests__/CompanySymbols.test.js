import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { useDispatch, useSelector } from 'react-redux';
import CompanySymbols from '../components/CompanySymbols';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

jest.mock('../redux/symbols/symbolsSlice', () => ({
  getSymbols: jest.fn(),
  getSearchResult: jest.fn(),
}));

describe('CompanySymbols', () => {
  beforeEach(() => {
    useDispatch.mockReturnValue(jest.fn());
    useSelector.mockReturnValue({
      slicedSymbols: [],
      symbols: [],
      error: undefined,
      hasMore: true,
    });
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Router>
          <CompanySymbols />
        </Router>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
