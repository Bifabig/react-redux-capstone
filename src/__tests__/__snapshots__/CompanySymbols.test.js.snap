// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompanySymbols renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="input"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        placeholder="Search..."
        type="text"
        value=""
      />
      <input
        type="submit"
        value="search"
      />
    </form>
  </div>
  <div
    className="stocks"
  >
    <h2>
      <i
        className="bi bi-bar-chart-steps"
      />
      Stocks
    </h2>
    <div
      className="symbols"
    >
      <div
        className="btnContainer"
      >
        <button
          className="load"
          onClick={[Function]}
          type="button"
        >
          Load More
        </button>
      </div>
    </div>
  </div>
</div>
`;
